proposed/http-message.md

```diff
@@ -393,11 +393,150 @@ Implementations are expected to:
   the tree.
 
+The tree structure referenced should mimic the naming structure in which files
+were submitted. 
+
+In the simplest example, this might be a single named form element submitted as:
+
+```html
+<input type="file" name="avatar" />
+```
+
+In this case, the structure in `$_FILES` would look like:
+
+```php
+array(
+    'avatar' => array(
+        'tmp_name' => 'phpUxcOty',
+        'name' => 'my-avatar.png',
+        'size' => 90996,
+        'type' => 'image/png',
+        'error' => 0,
+    ),
+)
+```
+
+The normalized form returned by `getUploadedFiles()` would be:
+
+```php
+array(
+    'avatar' => /* UploadedFileInterface instance */
+)
+```
+
+In the case of an input using array notation for the name:
+
+```html
+<input type="file" name="my-form[details][avatar]" />
+```
+
+`$_FILES` ends up looking like this:
+
+```php
+array(
+    'my-form' => array(
+        'details' => array(
+            'avatar' => array(
+                'tmp_name' => 'phpUxcOty',
+                'name' => 'my-avatar.png',
+                'size' => 90996,
+                'type' => 'image/png',
+                'error' => 0,
+            ),
+        ),
+    ),
+)
+```
+
+And the corresponding tree returned by `getUploadedFiles()` should be:
+
+```php
+array(
+    'my-form' => array(
+        'details' => array(
+            'avatar' => /* UploadedFileInterface instance */
+        ),
+    ),
+)
+```
+
+In some cases, you may specify an array of files:
+
+```html
+Upload an avatar: <input type="file" "name="my-form[details][avatars][]" />
+Upload an avatar: <input type="file" "name="my-form[details][avatars][]" />
+```
+
+(As an example, JavaScript controls might spawn additional file upload inputs to
+allow uploading multiple files at once.)
+
+In such a case, the specification implementation must aggregate all information
+related to the file at the given index. The reason is because `$_FILES` deviates
+from its normal structure in such cases:
+
+```php
+array(
+    'my-form' => array(
+        'details' => array(
+            'avatars' => array(
+                'tmp_name' => array(
+                    0 => '...',
+                    1 => '...',
+                    2 => '...',
+                ),
+                'name' => array(
+                    0 => '...',
+                    1 => '...',
+                    2 => '...',
+                ),
+                'size' => array(
+                    0 => '...',
+                    1 => '...',
+                    2 => '...',
+                ),
+                'type' => array(
+                    0 => '...',
+                    1 => '...',
+                    2 => '...',
+                ),
+                'error' => array(
+                    0 => '...',
+                    1 => '...',
+                    2 => '...',
+                ),
+            ),
+        ),
+    ),
+)
+```
+
+An implementation of this specification would convert the above to:
```

Comment:
They MUST do it that way; that's what we're specifying. The whole idea behind the `getUploadedFiles()` implementation is that any given leaf in the tree is an `UploadedFileInterface` instance, and that leaves can be retrieved using the same structure in which they were defined when uploaded.